File: 2022/day03/ruby/solution.rb
#!/usr/bin/env ruby
require 'set'

def parse_input(filename)
  File.readlines(filename).map(&:strip).reject(&:empty?)
end

def priority(char)
  if char.match?(/[a-z]/)
    char.ord - 'a'.ord + 1
  else
    char.ord - 'A'.ord + 27
  end
end

def part1(rucksacks)
  rucksacks.sum do |rucksack|
    mid = rucksack.length / 2
    first = rucksack[0...mid].chars.to_set
    second = rucksack[mid..].chars.to_set
    common = (first & second).first
    priority(common)
  end
end

def part2(rucksacks)
  rucksacks.each_slice(3).sum do |group|
    common = group.map { |r| r.chars.to_set }.reduce(:&).first
    priority(common)
  end
end

def main
  script_dir = File.dirname(File.expand_path(__FILE__))
  input_file = File.join(script_dir, '..', 'input.txt')

  rucksacks = parse_input(input_file)

  puts "Part 1: #{part1(rucksacks)}"
  puts "Part 2: #{part2(rucksacks)}"
end

main if __FILE__ == $PROGRAM_NAME

File: 2022/day03/php/solution.php
<?php

function parse_input($filename) {
    $lines = file($filename, FILE_IGNORE_NEW_LINES | FILE_SKIP_EMPTY_LINES);
    return $lines;
}

function priority($char) {
    if (ctype_lower($char)) {
        return ord($char) - ord('a') + 1;
    } else {
        return ord($char) - ord('A') + 27;
    }
}

function part1($rucksacks) {
    $total = 0;
    foreach ($rucksacks as $rucksack) {
        $mid = strlen($rucksack) / 2;
        $first = str_split(substr($rucksack, 0, $mid));
        $second = str_split(substr($rucksack, $mid));
        $common = array_intersect($first, $second);
        $total += priority(reset($common));
    }
    return $total;
}

function part2($rucksacks) {
    $total = 0;
    for ($i = 0; $i < count($rucksacks); $i += 3) {
        $first = str_split($rucksacks[$i]);
        $second = str_split($rucksacks[$i + 1]);
        $third = str_split($rucksacks[$i + 2]);
        $common = array_intersect($first, $second, $third);
        $total += priority(reset($common));
    }
    return $total;
}

function main() {
    $script_dir = dirname(__FILE__);
    $input_file = $script_dir . '/../input.txt';

    $rucksacks = parse_input($input_file);

    echo 'Part 1: ' . part1($rucksacks) . "\n";
    echo 'Part 2: ' . part2($rucksacks) . "\n";
}

main();
